### PR TITLE
Combine local + freeway profile requests for auto routing

### PR DESCRIPTION
**Pull request by mark-idleman**
Main idea laid out in this [thread](https://replicahq.slack.com/archives/CK358RL0Z/p1662666365540259). We _could_ in theory do this in the python router client (ask for routes for profile 1, then for profile 2, then combine), but this has 2 disadvantages:
- The `Route` python class we've made is a frozen dataclass, which means that it's not so simple to combine routes like this (they can't be modified after they're returned)
- The client-side approach results in 2x the number of grpc network calls during a run, which is much less optimal than keeping the number of network calls constant, and letting each call to the router simply handle 2x the number of routes internally (with this behavior hidden from view for the client)

What I did here:
- Take the two profiles Alexei posted [[link](https://replicahq.slack.com/archives/CK358RL0Z/p1662761354365509)], and add them as profiles used to build our production routers with names `car_local` and `car_freeway` (removing the old, basic `car` profile)
- Update the GTFS link mapper to use the `car_local` profile (equivalent to the `car` profile currently used in prod)
- Update the street routing server endpoint to route using both new car profiles + to combine the results of each request "under the hood" when a request specifies `car` as the mode (so, no change needed on the client side to specify something other than the standard `car`)

More specifically on the last point, we're now doing this for every street routing request:
- Given the "base" mode of the request coming from the user (eg `car`), find all profiles that our router has been built with that are prefixed by that mode (so for this example, `car_local` and `car_freeway`). Call this set of pre-built profiles `profilesToQuery`
- For each of `profilesToQuery`, perform a routing request with that profile. All other settings stay the same for each request made
- If _any_ of the requests throws an internal error (infrequent), the overall request fails, and sends a grpc `INTERNAL` error (as had happened before in the single request setup)
- If _none_ of the requests returns any results, return a grpc `NOT_FOUND` error, as had happened before if the single request found no results. NOTE: if at least 1 request that occurs returns valid paths, but other requests do not, no error is thrown - instead, whatever paths _were_ found are returned
- For all paths found over all requests, add the paths to the final response object and return it. This essentially adds the paths returned from multiple internal routing requests as "alternatives" returned with a single response object

Note that the above approach extends to any mode we'd like to use multiple profiles for, and works for any number `n >= 1` profiles you define, so long as they're prefixed by the "base" profile name (car, bike, etc)

It might be easiest to view the (important) changes to the street routing endpoint in unified view for easier comprehension [[link](https://github.com/replicahq/graphhopper/pull/116/files?diff=unified&w=0#diff-a792a98870506f2781021950718041253fd6c8d472ad13c969ef19baf9f210da)]

---

Tested in the most prod way possible 🙃 https://replicahq.slack.com/archives/CK358RL0Z/p1662913914035919

### CODE REVIEW

**Comment by bnaul** [comment]
probably not qualified to really review the java code but seems like a very creative and elegant approach to me, I agree that moving the "combine these profiles" stuff up into the service keeps things way way simpler.

is there still a "no alternatives" single-route option under this setup? I don't know that we ever even use it but would be good to know what happens in that case

**Comment by mark-idleman** [comment]
> is there still a "no alternatives" single-route option under this setup?

Really good question - there is and there isn't, lol. Whatever alt routing parameters you pass to the request get applied to every request in the `profilesToQuery` list. So, if you only define one car profile and request `car` with `alternatives=1`, you'll get 1 route, but if you've defined `n` car profiles, you'll get `n` routes 🤠 [this means the version we're testing w/ now returns up to 2 routes if you set `alternatives=1` and `mode=car`]

It's pretty easy to adjust that behavior if we want to, I guess (maybe just return the very first result only if `alternatives=1`? ie, making the first `car_<>` profile you define in your graphhopper config file effectively the "base" car profile for your setup). This whole change does mean that asking for as specific max alternative routes value isn't even really respected anymore, if you've defined >1 profile per mode

---

One other note related to this is the case where there are dupes in the combined set of routes we get back [[thread](https://replicahq.slack.com/archives/CK358RL0Z/p1662778705024249)]. Seems like a bridge we can cross later if we really want to

**Review by bryanwilliams1025 — COMMENTED** [review]
Nice - I like the concept behind this, and I like the idea of hiding this complexity from the client. But some parts of the API could become pretty unintuitive to clients with this change. You mentioned that `alternate_route_max_paths` in the request may not be respected anymore because we're finding alternates for each profile, but I also think `alternate_route_max_share_factor` may not be respected since an alternate for car_prof_1 could be identical to the best path for car_prof_2. There's a similar gotcha with `alternate_route_max_weight_factor`. I'm not sure how often we'd run into issues here in practice (at least currently with `car_local` vs. `car_freeway`), but it seems like this could eventually bite us. Still gaining context though so let me know if I'm off here 🙂

If we want to make this multi-profile behavior less surprising to clients, we could rename parts of the API (e.g. rename `profile` in `StreetRouteRequest` to `base_profile` or `profile_prefix` or something). We could also have each route in the response include the full profile name which created it. But this doesn't solve the confusing semantics around the alternate route request fields, and I'm not sure if we can/should do some postprocessing in RouterImpl so that those alternate route fields are respected.

Potentially stupid question - could we change `StreetRouteRequest` so the client can specify multiple profiles instead of just one? We could still have the alternate route fields apply on a per-profile basis, but the behavior should be less surprising to clients since they'd explicitly request multiple profiles. We could even reorganize the fields (or create a new type for multi-profile requests) so the client specifies `alternate_route_max_paths`, `alternate_route_max_share_factor`, and `alternate_route_max_share_factor` alongside each requested profile, making the per-profile alternate route behavior crystal clear.

File: test_export_gh_config.yaml [hunk 1]
```diff
@@ -3,25 +3,18 @@
 
 graphhopper:
   datareader.file: ./test-data/micro_nor_cal.osm.pbf
-  gtfs.max_transfer_interpolation_walk_time_seconds: 300
```
[the remark this answers is not in the document]
Comment (mark-idleman) [reply test_export_gh_config.yaml:6]: Good catch - yes, that's exactly right! I realized there was some unnecessary cruft in this config that shouldn't be there, so just removed it while it was still in my head

File: grpc/src/main/java/com/replica/RouterImpl.java [hunk 2]
```diff
@@ -92,85 +93,106 @@ public RouterImpl(GraphHopper graphHopper, PtRouter ptRouter, MatrixAPI matrixAP
...
+
+            PMap hints = new PMap();
+            hints.putObject(INSTRUCTIONS, false);
+            if (request.getAlternateRouteMaxPaths() > 1) {
+                ghRequest.setAlgorithm("alternative_route");
+                hints.putObject("alternative_route.max_paths", request.getAlternateRouteMaxPaths());
+                hints.putObject("alternative_route.max_weight_factor", request.getAlternateRouteMaxWeightFactor());
+                hints.putObject("alternative_route.max_share_factor", request.getAlternateRouteMaxShareFactor());
+            }
+            ghRequest.getHints().putAll(hints);
```
[the remark this answers is not in the document]
Comment (mark-idleman) [reply grpc/src/main/java/com/replica/RouterImpl.java:126]: Hmm yeah, I think that would work! Definitely cleans this up a bit. I'll get the refactor up in a few

**Comment by mark-idleman** [comment]
Thanks for the review @bryanwilliams1025 ! I'll try to hit on the main points you brought up here, happy to chat on a huddle if "in-person" follow-up would be helpful here because I know this is a context-heavy PR to get thrown into :) 

re: the alternate route knobs (all of them) not necessarily being respected anymore with this change, I 100% agree, we're not only losing the guarantee that the max paths specified by the client gets respected, but also that the "how different are the returned routes?" knobs won't have the same meaning when >1 profile get used vs. current behavior. Ultimately, this is a tradeoff we're forced to make (and human users of the client seem very willing to make 😉 ) - these knobs dictate behavior that's very low-level algorithmically (for each single request), so when combining the results of multiple requests, the whole notion of max share factor/weight factor kind of crumbles (ie we couldn't do some magic calculation after-the-fact that culls routes that don't respect the global settings passed by the client). We can even get routes that are 100% dupes (as mentioned above).

While we can (/should eventually) add _some_ basic follow-up logic that deals with this somewhat (getting rid of dupes, passing different knob settings per-profile, etc), I think our immediate-term comfort with the current approach is driven by the fact that getting more routes at all is generally better for us from a model quality perspective, and being super rigorous about the conceptual inconsistencies these knobs now have on the server might not have a large effect on that end product (or context, our chronic issue has been getting too few routes returned per request)

---

re: the potential for confusion on the client side due to this being hidden from variable names/not explicit when making a router call/etc, I generally agree with all of the points you've made - there are tweaks we can make here to be sure it's very apparent to the client + readers of the proto/server code exactly which profiles are getting queried, and how (setting different alt routing knobs per profile? etc etc). Your last few sentences are spot on - and in our plans for the near future! The main reasons I didn't touch this stuff here are
- We're trying to merge something in the middle of the official push to build the next season of Places, which means timing is critical (classic, I know...)
- As the idea of how exactly we want to enable python codebase experts to tweak/build/test router changes comes together, it's becoming clear that moving graphhopper configs over to the transportation network CDP (in the model repo) is the right call, and I think that will be a good time to refine the API/handling of multiple profiles-per-request once we've decided exactly how that will work (setting alt routing knobs next to their relevant profiles will be a lot easier here!). I see that essentially as a follow-up task to this ticket, after we can take a breath 
  - [having the configs in the model repo also helps reinforce to client users exactly what's getting queried]
  
edit: I guess option 2 is that I still merge [this](https://github.com/replicahq/model/pull/6112) to use the existing code immediately in Places ( 😈 ), and we keep this PR open a bit longer to give us time to figure out my last bullet point above heh. I don't have a strong preference

**Review by bryanwilliams1025 — APPROVED** [review]
Thanks for walking me through all that! Understand the urgency and the primary goal of getting more routes. I'm on board with merging this and cleaning things up once the GH configs move under the TN CDP, no need to block on these more academic concerns about knob consistency 🙂

**Pull request merged**